reframe/frontend/executors/policies.py

```diff
@@ -224,19 +262,49 @@ def runcase(self, case):
             self._failall(e)
             raise
 
+    def _cleanup_all(self):
+        for task in self._retired_tasks:
+            if task.ref_count == 0:
+                task.cleanup(not self.keep_stage_files)
+                self._retired_tasks.remove(task)
+
     def _poll_tasks(self):
         '''Update the counts of running checks per partition.'''
         getlogger().debug('updating counts for running test cases')
         getlogger().debug('polling %s task(s)' % len(self._running_tasks))
         for t in self._running_tasks:
             t.poll()
 
+    def _poll_waiting(self):
+        for partname, _ in self._running_tasks_counts.items():
+            for task in self._waiting_tasks[partname]:
```

Comment:
Why a double loop here? The following should be enough:

```python
for partname, task in self._waiting_tasks.items():
```

As suggested before, remove the tasks from the list in a separate list.